src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,316 @@
+
+            return GroupDocumentSymbolTrees(allSymbols)
+                .Select(group => CreateDocumentSymbolTree(group))
+                .ToArray();
+
+            // Groups a flat array of Document Symbols into arrays containing the symbols of each tree.
+            // The first symbol in an array is always the parent (determines the group's position range).
+            static ImmutableArray<ImmutableArray<DocumentSymbol>> GroupDocumentSymbolTrees(ImmutableArray<DocumentSymbol> allSymbols)
+            {
+                var documentSymbolGroups = ArrayBuilder<ImmutableArray<DocumentSymbol>>.GetInstance();
+                if (allSymbols.Length == 0)
+                {
+                    return documentSymbolGroups.ToImmutableAndFree();
+                }
+
+                var currentGroup = ArrayBuilder<DocumentSymbol>.GetInstance(1, allSymbols.First());
+                var currentRange = allSymbols.First().Range;
+                for (var i = 1; i < allSymbols.Length; i++)
+                {
+                    var symbol = allSymbols[i];
+                    // If the symbol's range is in the parent symbol's range
+                    if (symbol.Range.Start.Line > currentRange.Start.Line && symbol.Range.End.Line < currentRange.End.Line)
+                    {
+                        currentGroup.Add(symbol);
+                    }
+                    else
+                    {
+                        // Push existing group
+                        documentSymbolGroups.Add(currentGroup.ToImmutableAndFree());
+                        // Create new group with this symbol as the parent
+                        currentGroup = ArrayBuilder<DocumentSymbol>.GetInstance(1, symbol);
+                        currentRange = symbol.Range;
+                    }
+                }
+
+                documentSymbolGroups.Add(currentGroup.ToImmutableAndFree());
+                return documentSymbolGroups.ToImmutableAndFree();
+            }
+
+            // Given a flat array containing a Document Symbol and its descendants, returns the Document Symbol
+            // with its descendants recursively nested. The first Document Symbol in the array is considered the root node.
+            static DocumentSymbol CreateDocumentSymbolTree(ImmutableArray<DocumentSymbol> documentSymbols)
+            {
+                var node = documentSymbols.First();
+                var childDocumentSymbols = documentSymbols.RemoveAt(0);
+                node.Children = GroupDocumentSymbolTrees(childDocumentSymbols)
+                    .Select(group => CreateDocumentSymbolTree(group))
+                    .ToArray();
+                return node;
+            }
+        }
+
+        /// <summary>
+        /// Converts an array of type DocumentSymbol to an array of type DocumentSymbolItem.
+        /// </summary>
+        public static ImmutableArray<DocumentSymbolItem> GetDocumentSymbolItems(DocumentSymbol[] documentSymbols)
```

Comment:
nit: because DocumentSYmbolItem and DocumentSymbol are so close in naming, consider adding hte following to thsi file `using LspDocumentSymbol = DocumentSymbol;`.  And then use `LspDocumentSymbol` in the file instead.  this will make it clearer as to which domains you're moving between.  right now it's easy to get confused as these names are so similar.

We do the same with things like `Completion` and `VSCompletion` to make it more readily apparent which we are referring to.